libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -786,42 +790,81 @@ void _serializeConnect( const IotMqttConnectInfo_t * pConnectInfo,
     }
 
     /* If metrics are enabled, write the metrics username into the CONNECT packet.
-     * Otherwise, write the username and password only when not connecting to an
+     * Otherwise, write the username and password only when not connecting to the
      * AWS IoT MQTT server. */
     if( pConnectInfo->awsIotMqttMode == true )
     {
         #if AWS_IOT_MQTT_ENABLE_METRICS == 1
             IotLogInfo( "Anonymous metrics (SDK language, SDK version) will be provided to AWS IoT. "
                         "Recompile with AWS_IOT_MQTT_ENABLE_METRICS set to 0 to disable." );
 
-            pBuffer = _encodeString( pBuffer,
-                                     AWS_IOT_METRICS_USERNAME,
-                                     AWS_IOT_METRICS_USERNAME_LENGTH );
-        #endif
+            /* Determine if the Connect packet should use a combination of the username
+             * for authentication plus the SDK version string. */
+            if( pConnectInfo->pUserName != NULL )
+            {
+                /* Only include metrics if it will fit within the encoding
+                 * standard. */
+                if( ( pConnectInfo->userNameLength + AWS_IOT_METRICS_USERNAME_LENGTH ) <= UINT16_MAX )
```

Comment:
This check should be moved to _connectPacketSize.  It could potentially fail silently here and generate an invalid MQTT packet.